### ISSUE TITLE
I want to change the buildDictionary in HasOneOrMany.

### ISSUE THREAD

**Issue by hrs-o** [laravel/ideas#2369]
Hi, I have a slightly different use of model properties in my project.
For example, we are making use of a mutator property with the same name as a model property, as shown below.

```php
/**
 * @property int $user_id
 */
class User
{
    public function getUserIdAttribute(): string
    {
        return obfuscation('user_id');
    }

    public function getOriginalUserIdAttribute(): int
    {
        return $this->getOriginal('user_id');
    }
}

```

This behavior in itself is satisfactory, but there was a problem when I tried to relate it to another table.
This is due to the fact that `$result->{$foreign}` in HasOneOrMany#buildDictionary, so it is obtained as a mutator.
As a result, HasOneOrMany#matchOneOrMany, which refers to the actual data, cannot get the data as a relation.

To fix this, I would like to modify `buildDictionary` for the following.

```diff
    protected function buildDictionary(Collection $results)
    {
        $foreign = $this->getForeignKeyName();

-        return $results->mapToDictionary(function ($result) use ($foreign) {
-            return [$result->{$foreign} => $result];
+        return $results->mapToDictionary(
+            fn($result) => [$result->getOriginal($foreign) => $result]
        )->all();
    }
```

In summary, I changed `$result->{$foreign}` to `$result->getOriginal($foreign)`.

ref.
https://github.com/laravel/framework/blob/64f072cc6f106d6871a0a25c537fac48c6e6df39/src/Illuminate/Database/Eloquent/Relations/HasOneOrMany.php#L165
